composeApp/src/commonMain/kotlin/sp/bvantur/tasky/login/presentation/LoginViewModel.kt

```diff
@@ -0,0 +1,73 @@
+package sp.bvantur.tasky.login.presentation
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.launch
+import sp.bvantur.tasky.core.DispatcherProvider
+import sp.bvantur.tasky.core.domain.ValidateEmailUseCase
+import sp.bvantur.tasky.core.domain.ValidatePasswordUseCase
+import sp.bvantur.tasky.core.presentation.ViewModelUserActionHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandlerImpl
+
+class LoginViewModel(
+    private val dispatcherProvider: DispatcherProvider,
+    private val validateEmailUseCase: ValidateEmailUseCase,
+    private val validatePasswordUseCase: ValidatePasswordUseCase
+) : ViewModel(),
+    ViewModelUserActionHandler<LoginUserAction>,
+    ViewModelViewStateHandler<LoginViewState> by ViewModelViewStateHandlerImpl(
+        LoginViewState(),
+        dispatcherProvider
+    ) {
+    override fun onUserAction(userAction: LoginUserAction) {
+        when (userAction) {
+            is LoginUserAction.EmailChanged -> onEmailChanged(userAction.value)
+            is LoginUserAction.PasswordChanged -> onPasswordChanged(userAction.value)
+            LoginUserAction.OnLogin -> onLogin()
+        }
+    }
+
+    private fun onLogin() {
+        viewModelScope.launch {
+            val isEmailValid = validateEmailUseCase.invoke(viewStateFlow.value.email)
+            val isPasswordValid = validateEmailUseCase.invoke(viewStateFlow.value.email)
+
+            if (!isEmailValid || !isPasswordValid) {
+                emitViewState(
+                    viewStateFlow.value.copy(
+                        isEmailError = !isEmailValid,
+                        isPasswordError = !isPasswordValid
+                    )
+                )
+                return@launch
+            }
+
+            // TODO implement backend communication
+        }
+    }
+
+    private fun onEmailChanged(value: String) {
+        viewModelScope.launch {
+            val isValid = validateEmailUseCase.invoke(value)
+            emitViewState(
+                viewStateFlow.value.copy(
+                    isEmailError = !isValid,
+                    email = value
+                )
+            )
+        }
+    }
+
+    private fun onPasswordChanged(value: String) {
+        viewModelScope.launch {
+            val isValid = validatePasswordUseCase.invoke(value)
+            emitViewState(
+                viewStateFlow.value.copy(
+                    isPasswordError = !isValid,
+                    password = value
+                )
+            )
```

Comment:
I wonder, if this could be vulnerable to race conditions. You are making use of `update` in your emit function, but there's a delay between calling `copy` and actually updating the state with it in which a race condition could technically be possible if I'm not wrong 🤔 
Normally, the `update` function gives you the most-recent state as a lambda argument which you can use:
```
_state.update {
    it.copy(...) 
}
```
I think it's very unlikely, since you also switch to the main thread where race conditions can't happen, but nothing guarantees that the `copy` call is also called on the main thread. In the time the dispatcher takes to switch, the state could've been changed to something else already, but after switching you'd update it with the previous copy.

You should be able to fix that by providing a lambda to update the state instead, similar as `update` does it. In that case, the call to copy would happen inside the `update` block and not at the moment where you call `emitViewState`.